tests/kernel/smp/src/main.c

```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
+	k_sem_init(&sema, 0, 1);
+
+	tinfo[0].tid =
+	k_thread_create(&tthread[0], tstack[0], STACK_SIZE,
+			(k_thread_entry_t)t_lock_unlock,
+			&smutex, NULL, NULL,
+			K_PRIO_PREEMPT(1),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	tinfo[1].tid =
+	k_thread_create(&tthread[1], tstack[1], STACK_SIZE,
+			(k_thread_entry_t)t2_do_irq,
+			&smutex, NULL, NULL,
+			K_PRIO_PREEMPT(1),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_sem_take(&sema, K_FOREVER);
+
+	zassert_equal(interrupt_happened[0], 0,
+		"invalid initial int count(%d)", interrupt_happened[0]);
+
+	irq_offload((irq_offload_routine_t)tIsr_check_entry, (void *)0);
+
+	/* check if interrupt still works */
+	zassert_equal(interrupt_happened[0], 1,
+		"invalid initial int count(%d)", interrupt_happened[0]);
+
+	k_sem_give(&sema);
+
+	k_thread_join(tinfo[0].tid, K_FOREVER);
+	k_thread_join(tinfo[1].tid, K_FOREVER);
+	cleanup_resources();
+}
+
+static void t1_mutex_lock(void *p1, void *p2, void *p3)
+{
+	/* t1 will get mutex first */
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	k_msleep(2);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+}
+
+static void t2_mutex_lock(void *p1, void *p2, void *p3)
+{
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+
+	/**TESTPOINT: z_smp_release_global_lock() has been call during
+	 * context switch but global_lock_cnt has not been decrease
+	 * because no irq_lock() was called.
+	 */
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+}
+
+static void t2_mutex_lock_with_irq(void *p1, void *p2, void *p3)
+{
+	int key;
+
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	key = irq_lock();
+
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	zassert_equal(_current->base.global_lock_count, 1,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+
+	/**TESTPOINT: Though the irq has not been locked yet, but the
+	 * global_lock_cnt has been decreased during context switch.
+	 */
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	irq_unlock(key);
+
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+}
+
+/**
+ * @brief Test scenairo that a thread release the global lock
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate the scenario that make the internal APIs of SMP
+ * z_smp_release_global_lock() to be called.
```

Comment:
Seems like there should be some way to exercise this with just plain irq_lock/unlock().  But whiteboxing this isn't awful, it's a complicated internal state and an exerciser with the public API would be no less confusing.

Reply:
Yes, got your point. I believe that there is some better way that uses less code but I spent lots of this on this without design the better testcase to cover it. I will keep finding them out. So can I just keep them at this time? Because they can cover all the branches of the function z_smp_release_global_lock(). Thank you!